[SCRUM-60] ✨ User 관련 API 추가

trivialcoding6 opened the pull request:
## Explain this Pull Request 🙏

- Jira URL: https://9git-9git.atlassian.net/jira/software/projects/SCRUM/issues/SCRUM-60?jql=project%20%3D%20%22SCRUM%22%20ORDER%20BY%20created%20DESC

## What has fixed? 🛠

변경 전 내용

변경 후 내용
- User 관련 API 추가
- 공통 로직 추가
* Response 형태 통일: https://github.com/9git9git/9git-backend/commit/55754b966da6329b6d2c8ae243dda049dfed56d8
* CORS 설정 추가: https://github.com/9git9git/9git-backend/commit/abbea4b5532f5a4c52d215091ce68708a74364e4

추가로 https://github.com/9git9git/9git-backend/commit/7cf2e97c707f1684f192bbaec325660b615783d8 여기 PR에서 DB 수정이 필요합니다. 테스트를 해보시려면 @trivialcoding6 에게 문의해주시면 테스트 가능한 DB 연결 정보 드리겠습니다.
테스트는 ```poetry run``` 이후 http://localhost:8000/docs 에서 테스트 가능합니다.

## Whether needed review 🖍

<!-- 마크다운의 빈 체크박스 문법은 [ ] / 채워진 체크박스 문법은 [x] 입니다. -->

- [x] 🙋🏻 리뷰가 필요합니다!
- [ ] 🙅🏻 리뷰가 필요하지 않습니다!

(이유 : 기능 추가)

## Document🗒️ (Optional)

## Screenshot 📸 (Optional)

## Test Checklist ✅ (Optional)

<!-- 마크다운의 빈 체크박스 문법은 [ ] / 채워진 체크박스 문법은 [x] 입니다. -->

trivialcoding6 committed:
🐛 Db복합키 관련 에러 해결

동준님이 먼저 올려주신 commit 보고 작업했습니다.

trivialcoding6 committed:
🐛 User 테이블에 optional 값 명시적으로 type hint 제공

optional 값을 명시적으로 제공하지 않으면 필수값으로 처리됨.

trivialcoding6 committed:
✨ Base 및 user schema 추가

base에는 모든 스키마에 해당하는 from_attributes 와 camel_case로 변환하는 로직을 작성해놓았습니다.

trivialcoding6 committed:
✨ User cRUD 추가

trivialcoding6 committed:
✨ User Service 추가

trivialcoding6 committed:
✨ User API endpoints 추가

trivialcoding6 committed:
✨ Response 형태 통일

API Response 형태를 프론트에서 사용하기 편리하게 통일시켰습니다.

trivialcoding6 committed:
✨ CORS 설정 추가

프론트에서 API 요청이 가능하게 CORS 허용 설정 추가

GideokKim commented on app/models/evaluation.py:
```diff
@@ -50,31 +47,25 @@ class RecommendedChallenge(Base):
     )
+    category_progress: Mapped["CategoryProgress"] = relationship(
+        "CategoryProgress", back_populates="comprehensive_evaluations"
+    )
```
`relationship` 관련 코드들 위치가 변경된 부분들이 좀 있는데, 특별한 이유가 있을까요? 내용이 변경된 것 같지는 않아서 여쭤봅니당

trivialcoding6 replied:
@GideokKim 
엇 그렇네요 원위치 시켜놨습니다!
https://github.com/9git9git/9git-backend/pull/6/commits/5b58e735f77a0a4bf523a12e72387365a798c130

GideokKim commented on app/models/evaluation.py:
```diff
@@ -94,26 +85,21 @@ class ComprehensiveEvaluation(Base):
+    )
+    category_progress: Mapped["CategoryProgress"] = relationship(
+        "CategoryProgress", back_populates="monthly_achievements"
+    )
```
여기도 마찬가지로 위치 변경에 대해 특별한 이유가 있으신지 여쭤봅니당

trivialcoding6 replied:
@GideokKim 
여기도 동일하게 원위치 시켰습니다!

GideokKim commented on app/models/user.py:
```diff
@@ -18,50 +18,52 @@ class User(Base):
-    level: Mapped[int] = mapped_column(Integer, default=1)
-    exp: Mapped[int] = mapped_column(Integer, default=0)
-    character_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
+    password: Mapped[str] = mapped_column(String(255))
```
여기 `nullable=False`가 빠졌는데, 특별한 이유가 있으실까용? 사실 패스워드를 빼먹을 일은 없긴하지만 여쭤봅니당 ㅎㅎ

trivialcoding6 replied:
@GideokKim 
엇 위에꺼 optional로 바꾸다가 자동 완성으로 같이 바꿔진거 같아요 수정했습니다!
https://github.com/9git9git/9git-backend/pull/6/commits/a4a33f3e9b964328267747368d726376af21754e

GideokKim commented on app/models/user.py:
```diff
@@ -18,50 +18,52 @@ class User(Base):
 
     # User → UserCharacter, Goal, TodayNote, CategoryProgress, Chat, Storage, ComprehensiveEvaluation, MonthlyAchievement (1:N 관계)
-    user_characters: Mapped[List["UserCharacter"]] = relationship(
+    user_characters: Mapped[Optional[List["UserCharacter"]]] = relationship(
```
관계에 대한 `Optional`은  어떤 이유로 필요한건가요? `Optional` 값을 명시적으로 제공하지 않으면 필수값으로 처리된다고 하셨는데, 관계 설정은 필수적인 부분이라 생각해서용

trivialcoding6 replied:
@GideokKim 
아 네네 관계는 기본적으로는 필수값이 아니라고 생각합니다! 필수값이 된다고 한다면 유저를 만들때(회원가입할때) 모든 관계에 대한 데이터도 같이 추가를 해줘야합니다! 목표, 오늘의 노트 등등 저희는 처음 가입할 때 해당 내용들이 없을 수 있기 때문에 Optional로 했습니다. 유저 캐릭터 역시 기본 캐릭터가 있나 싶은데 chracter_counter가 defalut=0 이라 이 컬럼 역시도 필수값은 아닌 것으로 보입니다.

GideokKim replied:
넵 설명 감사합니다 ㅎㅎ resolve 처리하겠습니당

GideokKim commented on app/schemas/user.py:
```diff
@@ -0,0 +1,52 @@
+from app.schemas.base import BaseModel
+
+
+class GenderEnum(str, Enum):
+    MALE = "M"
+    FEMALE = "F"
```
저희 models에는 GenderEnum을 `PyEnum`을 사용했는데, 여기는 `Enum`을 사용했네요! 혹시 달라도 괜찮을까요?

그리고 crud쪽에서는 models에 있던 GenderEnum을 import해서 썼는데, 여기서는 그렇게 사용하면 안되는지에 대해서도 궁금합니당

Ryan-OH replied:
저도 이부분 궁금해요

trivialcoding6 replied:
@GideokKim @Ryan-OH 
모델에서 사용하는 GenderEnum 역시 동일한 Enum 입니다! as 를 통해 PyEnum 으로 별칭을 붙여 사용중인것으로 보입니다.
<img width="553" alt="스크린샷 2025-04-10 오전 10 40 04" src="https://github.com/user-attachments/assets/36bd6842-b37d-45fb-8acf-5668ed93d1a3" />

관련해서 Enum 같은 경우는 이번처럼 같은게 여러 곳에서 사용될 거 같은데 enum 폴더를 하나 만들어서 관리를 하는게 나을까요?

Ryan-OH replied:
아 제 질문은 모델부분에서 GenderEnum임포트해서 사용하지 않는 이유가 있을까 싶어서 질문드렸습니다

GideokKim replied:
@trivialcoding6  저는 import해서 쓴다고 하면  따로 모아두면 좋을 것 같긴합니당

trivialcoding6 replied:
@GideokKim @Ryan-OH 
그럼 폴더 만들고 적절히 파일을 만들어서.... 올려볼게요! enum은 어떻게 파일을 묶을지도 고민이네요 일단 제가 고민해보고 올려볼테니 그때 리뷰 한번 더 부탁드립니다!

GideokKim replied:
확인 해봤을 때 너무 잘게 쪼개놨나싶은 생각도 들지만 이대로 써도 문제는 없고 추후 리펙터링해야 할때나 다시 생각해보면 좋을 것 같아서  resolve 처리하겠습니당

GideokKim commented on app/schemas/user.py:
```diff
@@ -0,0 +1,52 @@
+    id: UUID
+    user_id: UUID
+    character_id: UUID
+    character: Optional[CharacterResponse] = None
+    user: Optional[UserResponse] = None
```
이 부분은 관계된 테이블을 나타내주는 코드인가요?

trivialcoding6 replied:
@GideokKim 
아 네네!! 원래 유저랑 캐릭터 둘다하려다가 jira 이슈가 user 관련이라 character는 스키마만 만들고 작업을 안했습니다! 요 부분은 이 pr 머지되고 캐릭터쪽 작업하면서 다듬어 나갈거 같습니다!

GideokKim commented on app/crud/user.py:
```diff
@@ -0,0 +1,85 @@
+
+async def create_user(db: AsyncSession, user_data: UserCreate) -> UserResponse:
+
+    if await read_user_by_email(db, user_data.email):
+        raise HTTPException(status_code=400, detail="이미 존재하는 이메일입니다.")
```
혹시 이 함수 때문에 앞에 read 함수들부터 작성해야 하는건가용?

crud라서 create 함수들이 먼저 작성될줄 알았는데 순서가 read부터라서 궁금해서 여쭤봅니당

trivialcoding6 replied:
@GideokKim 
아 순서는 크게 의미를 두진 않았고 read 부터 개발을 해서 자연스럽게 위로 왔습니다!
요거 맞추는게 나을까요?

trivialcoding6 replied:
@GideokKim 
일단 `read_user_by_email` 을 내렸을 때 에러는 따로 안나는거 같긴합니닷

GideokKim replied:
저는 다들 CRUD라고 해서 이 순서대로 작성하겠구나싶었어용 혹시 백엔드 개발하시는 분들의 관행적 순서가 있다면 그걸 따르는게 좋을 것 같고 그런게 없다면 개인적으로는 CRUD니까 CRUD 순서로 배치하면 좋지 않을까 싶은 생각입니당 ㅎㅎ(에러가 발생하지 않는다면!)

trivialcoding6 replied:
@GideokKim 
넵!! 그럼 위에사항이랑 요거 반영해서 작업해놓겠습니다!

GideokKim commented on app/api/v1/endpoints/user.py:
```diff
@@ -0,0 +1,71 @@
+        raise HTTPException(status_code=500, detail=str(e))
+
+
+@router.get("/email/{email}", response_model=UserResponse)
```
email은 `/email/{email}`로 작성하셨는데, user id에 대해서는 `/{user_id}`로 통일하셨더라구요. 혹시 다르게 해야 하는 이유가 있을까용?

trivialcoding6 replied:
@GideokKim 
REST API 설계할때 보통 id 라고 하면 상위 계층/{id}로 가서 현재 url이 /api/v1/users 가 상위 계층이므로 id는 그 뒤에 바로 붙여주었고, email은 따로 추가해주었습니다. 근데 정확하게 하려면 /emails/{email} 이 맞아보이긴 하네요. 이렇게 수정할까요?
![스크린샷 2025-04-10 오전 10 43 28](https://github.com/user-attachments/assets/7f3adad7-45c3-47c1-836c-5396aa37472a)

GideokKim replied:
` /emails/{email}`로 수정하면 좋을 것 같습니당 ㅎㅎ 설명 감사합니다!

trivialcoding6 replied:
@GideokKim 
반영 했습니다!
https://github.com/9git9git/9git-backend/pull/6/commits/9521c4f821b1f9c5ca0b5c8a3faade64133e9d45

GideokKim reviewed (commented):
리뷰 코멘트 남겼습니다!

Ryan-OH commented on app/schemas/user.py:
```diff
@@ -0,0 +1,52 @@
+class GenderEnum(str, Enum):
+    MALE = "M"
+    FEMALE = "F"
+
```
혹시 Enum따로 다시 해주신 이유가 있을까요? 저는 모델에 있는거 임포트 해주긴했는데 궁금해요!

trivialcoding6 committed:
🩹 리뷰반영: category 선언위치 제자리로

trivialcoding6 committed:
🩹 리뷰반영: 비밀번호 nullable=false 추가

GideokKim commented on app/models/evaluation.py:
```diff
@@ -50,31 +47,22 @@ class RecommendedChallenge(Base):
         ForeignKey("category_progresses.id"), nullable=False
     )
+
+    user_id: Mapped[UUID] = mapped_column(ForeignKey("users.id"), nullable=False)
```
코드 위치가 변경이 되었는데, 특별한 이유가 있을까용?

trivialcoding6 replied:
@GideokKim 
원위치로 수정했습니다!
https://github.com/9git9git/9git-backend/pull/6/commits/6ac000e917cd0ac5f00ca9dbfa36c34094ec9319

trivialcoding6 committed:
🩹 Enum 폴더 생성 후 enum 추가

trivialcoding6 committed:
🐛 함수 순서 변경 및 버그 수정

HTTPException 말고 일반 Exception 도 잡으니 데이터는 추가 되지만 에러가 나고 있었어서 관련 내용 수정했고, enum 통일화 하면서 기존의 enum 변경 로직이 필요 없어져 삭제 했습니다

trivialcoding6 committed:
🩹 로거 삭제

trivialcoding6 committed:
🩹 Email endpoints 수정

email/{email} -> emails/{email}

trivialcoding6 commented:
@GideokKim @Ryan-OH @chaeengg @codingorijazz 
https://github.com/9git9git/9git-backend/pull/6/commits/efc614f04cafd6fd419364efcc1004e23addfb42
그 endpoint에서 HTTPException 말고 일반 Exception 도 캐치하도록 했는데 보니까 에러나는 부분이 있어 수정했습니다
`db.add()` -> 비동기 아니라서 await 삭제
enum 을 빼서 공통으로 가져다 사용하면서 update 에서 enum 형변환이 필요 없어져 삭제했습니다

GideokKim commented on app/models/evaluation.py:
```diff
@@ -124,14 +95,8 @@ class MonthlyAchievement(Base):
+
     user: Mapped["User"] = relationship("User", back_populates="monthly_achievements")
-    category_progress: Mapped[List["CategoryProgress"]] = relationship(
+    category_progress: Mapped["CategoryProgress"] = relationship(
```
이거 관계가 변한것 같은데 ERD에서 변경된걸까용?

Ryan-OH replied:
저거 아마 N:1 관계에서 잘못수정되어있어서 바꾸신것 같아요 저도 저부분에러나서 동일하게 바꿧어요

trivialcoding6 replied:
@GideokKim 
네네 저도 오류나서 수정했었어요!!

GideokKim replied:
@trivialcoding6  @Ryan-OH   넵 알겠습니당 감사합니다!

GideokKim reviewed (commented):
추가 코멘트드립니당

trivialcoding6 committed:
🩹 리뷰 반영: 코드 원위치

GideokKim reviewed (approved):
LGTM

Ryan-OH reviewed (approved):
LGTM
